// ng-number-picker/esm2015/lib/number-picker.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NumberPickerComponent } from './number-picker.component';
import { CommonModule } from '@angular/common';
import { NumberPickerService } from './number-picker.service';
export class NumberPickerModule {
}
NumberPickerModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule
                ],
                providers: [NumberPickerService],
                declarations: [NumberPickerComponent],
                exports: [NumberPickerComponent]
            },] },
];
function NumberPickerModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NumberPickerModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NumberPickerModule.ctorParameters;
}
