// ng-number-picker/bundles/ng-number-picker.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ng-number-picker', ['exports', '@angular/core', '@angular/forms', '@angular/common'], factory) :
    (factory((global['ng-number-picker'] = {}),global.ng.core,global.ng.forms,global.ng.common));
}(this, (function (exports,core,forms,common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NumberPickerService = (function () {
        function NumberPickerService() {
            /**
             * Min picker value
             */
            this.min = 0;
            /**
             * Max picker value
             */
            this.max = 100;
            /**
             * Pick step value
             */
            this.step = 1;
            /**
             * Delay for start picking values
             */
            this.pickStartAfter = 500;
            /**
             * Delay betweens each pick
             */
            this.pickTimer = 100;
            /**
             * value precision
             */
            this.precision = 1;
            /**
             * Inital picker value
             */
            this.value = null;
        }
        NumberPickerService.decorators = [
            { type: core.Injectable },
        ];
        return NumberPickerService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NumberPickerComponent$$1 = (function () {
        function NumberPickerComponent$$1(numberPickerService) {
            this.numberPickerService = numberPickerService;
            this.eventHolder = null;
            this.countInterval = null;
            this.isInputFocused = false;
            this.size = 'md';
            this.customClass = {};
            this.mouseWheel = false;
            this.arrowKeys = true;
            this.inputReadOnly = false;
            this.showUpButton = true;
            this.showDownButton = true;
            this.valueChange = new core.EventEmitter();
            this.minReached = new core.EventEmitter();
            this.maxReached = new core.EventEmitter();
            this.pickStarted = new core.EventEmitter();
            this.pickStoped = new core.EventEmitter();
            this.pickUpStarted = new core.EventEmitter();
            this.pickUpStoped = new core.EventEmitter();
            this.pickDownStarted = new core.EventEmitter();
            this.pickDownStoped = new core.EventEmitter();
        }
        /**
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.initPicker();
            };
        /**
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.isHorizontal = /**
         * @return {?}
         */
            function () {
                return (this.buttonsOrientation !== 'v' && this.buttonsOrientation !== 'vertical');
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onFocus = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                event.preventDefault();
                event.stopPropagation();
                this.isInputFocused = true;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onBlur = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                event.preventDefault();
                event.stopPropagation();
                this.isInputFocused = false;
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onMouseWheel = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.isInputFocused) {
                    event.preventDefault();
                    var /** @type {?} */ wheelUp = null;
                    var /** @type {?} */ delta = null;
                    if (event.wheelDelta) {
                        delta = event.wheelDelta / 60;
                    }
                    else if (event.detail) {
                        delta = -event.detail / 2;
                    }
                    if (delta !== null) {
                        wheelUp = delta > 0;
                    }
                    this.afterMouseWheels(wheelUp, event);
                    event.stopPropagation();
                }
            };
        /**
         * @param {?} wheelUp
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.afterMouseWheels = /**
         * @param {?} wheelUp
         * @param {?} event
         * @return {?}
         */
            function (wheelUp, event) {
                this.onPickStarted(wheelUp);
                if (wheelUp) {
                    this.onIncrease(event);
                }
                else {
                    this.onDecrease(event);
                }
                this.onPickStoped(wheelUp);
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onValueChange = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.value > this.max) {
                    this.value = this.max;
                }
                else if (this.value < this.min) {
                    this.value = this.min;
                }
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onDecrease = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                event.preventDefault();
                if (this.canDecrease()) {
                    this.value = this.round((this.value > this.min) ? this.value -= this.step : this.value);
                    this.valueChange.emit(this.value);
                }
                else {
                    this.minReached.emit(true);
                }
                event.stopPropagation();
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onIncrease = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                event.preventDefault();
                if (this.canIncrease()) {
                    this.value = this.round((this.value < this.max) ? this.value += this.step : this.value);
                    this.valueChange.emit(this.value);
                }
                else {
                    this.maxReached.emit(true);
                }
                event.stopPropagation();
            };
        /**
         * @param {?} event
         * @param {?=} increase
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onMouseDown = /**
         * @param {?} event
         * @param {?=} increase
         * @return {?}
         */
            function (event, increase) {
                if (increase === void 0) {
                    increase = true;
                }
                this.afterMouseDown(increase, event);
            };
        /**
         * @param {?} keyCode
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.isArrowUpDown = /**
         * @param {?} keyCode
         * @return {?}
         */
            function (keyCode) {
                return keyCode === 38 || keyCode === 40;
            };
        /**
         * @param {?} keyCode
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.isArowUp = /**
         * @param {?} keyCode
         * @return {?}
         */
            function (keyCode) {
                return keyCode === 38;
            };
        /**
         * @param {?} increase
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.loopPick = /**
         * @param {?} increase
         * @param {?} event
         * @return {?}
         */
            function (increase, event) {
                var _this = this;
                this.onPickStarted(increase);
                this.eventHolder = setTimeout(function () {
                    _this.countInterval = setInterval(function () {
                        if (increase) {
                            _this.onIncrease(event);
                        }
                        else {
                            _this.onDecrease(event);
                        }
                    }, _this.pickTimer);
                }, this.pickStartAfter);
            };
        /**
         * @param {?} event
         * @param {?=} increase
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onMouseUp = /**
         * @param {?} event
         * @param {?=} increase
         * @return {?}
         */
            function (event, increase) {
                if (increase === void 0) {
                    increase = true;
                }
                this.afterMouseUp(increase, event);
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onKeyDown = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.isArrowUpDown(event.keyCode)) {
                    event.preventDefault();
                    if (!this.eventHolder) {
                        this.loopPick(this.isArowUp(event.keyCode), event);
                    }
                }
                event.stopPropagation();
            };
        /**
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onKeyUp = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                if (this.isArrowUpDown(event.keyCode)) {
                    event.preventDefault();
                    this.afterPick(this.isArowUp(event.keyCode));
                }
                event.stopPropagation();
            };
        /**
         * @param {?} up
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.afterMouseDown = /**
         * @param {?} up
         * @param {?} event
         * @return {?}
         */
            function (up, event) {
                event.preventDefault();
                if (this.isLeftClick(event.which)) {
                    this.loopPick(up, event);
                }
                event.stopPropagation();
            };
        /**
         * @param {?} up
         * @param {?} event
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.afterMouseUp = /**
         * @param {?} up
         * @param {?} event
         * @return {?}
         */
            function (up, event) {
                event.preventDefault();
                this.afterPick(up);
                event.stopPropagation();
            };
        /**
         * @param {?} up
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.afterPick = /**
         * @param {?} up
         * @return {?}
         */
            function (up) {
                this.onPickStoped(up);
                this.clearTimers();
            };
        /**
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.clearTimers = /**
         * @return {?}
         */
            function () {
                clearTimeout(this.eventHolder);
                clearInterval(this.countInterval);
                this.eventHolder = null;
                this.countInterval = null;
            };
        /**
         * @param {?} up
         * @param {?} event
         * @param {?} start
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.afterArrowKeysPressed = /**
         * @param {?} up
         * @param {?} event
         * @param {?} start
         * @return {?}
         */
            function (up, event, start) {
                if (start) {
                    this.loopPick(up, event);
                }
                else {
                    this.afterPick(up);
                }
            };
        /**
         * @param {?} value
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.parseVal = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                if (typeof value === 'number') {
                    return value;
                }
                return parseFloat(value);
            };
        /**
         * @param {?} step
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.getPrecision = /**
         * @param {?} step
         * @return {?}
         */
            function (step) {
                return /\d*$/.exec(String(step))[0].length;
            };
        /**
         * @param {?} value
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.round = /**
         * @param {?} value
         * @return {?}
         */
            function (value) {
                return Math.round(value * Math.pow(10, this.precision)) / Math.pow(10, this.precision);
            };
        /**
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.canIncrease = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ canIncrease = (this.value <= this.max - this.step);
                if (!canIncrease) {
                    this.value = this.max;
                }
                return canIncrease;
            };
        /**
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.canDecrease = /**
         * @return {?}
         */
            function () {
                var /** @type {?} */ canDecrease = (this.value >= this.min + this.step);
                if (!canDecrease) {
                    this.value = this.min;
                }
                return canDecrease;
            };
        /**
         * @param {?} increase
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onPickStarted = /**
         * @param {?} increase
         * @return {?}
         */
            function (increase) {
                if (increase) {
                    if (this.canIncrease()) {
                        this.pickStarted.emit(true);
                        this.pickUpStarted.emit(true);
                    }
                }
                else {
                    if (this.canDecrease()) {
                        this.pickStarted.emit(true);
                        this.pickDownStarted.emit(true);
                    }
                }
            };
        /**
         * @param {?} increase
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.onPickStoped = /**
         * @param {?} increase
         * @return {?}
         */
            function (increase) {
                if (increase) {
                    if (this.canIncrease()) {
                        this.pickUpStoped.emit(true);
                        this.pickStoped.emit(true);
                    }
                }
                else {
                    if (this.canDecrease()) {
                        this.pickDownStoped.emit(true);
                        this.pickStoped.emit(true);
                    }
                }
            };
        /**
         * @param {?} witch
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.isLeftClick = /**
         * @param {?} witch
         * @return {?}
         */
            function (witch) {
                return witch === 1;
            };
        /**
         * @return {?}
         */
        NumberPickerComponent$$1.prototype.initPicker = /**
         * @return {?}
         */
            function () {
                this.min = this.parseVal(this.min) || this.numberPickerService.min;
                this.max = this.parseVal(this.max) || this.numberPickerService.max;
                this.step = this.parseVal(this.step) || this.numberPickerService.step;
                this.value = this.parseVal(this.value) || this.numberPickerService.value;
                this.pickStartAfter = this.parseVal(this.pickStartAfter) || this.numberPickerService.pickStartAfter;
                this.pickTimer = this.parseVal(this.pickTimer) || this.numberPickerService.pickTimer;
                this.precision = this.getPrecision(this.step) || this.numberPickerService.precision;
                this.value = this.round(this.value);
                if (this.value < this.min) {
                    this.value = this.min;
                }
                else if (this.value > this.max) {
                    this.value = this.max;
                }
            };
        NumberPickerComponent$$1.decorators = [
            { type: core.Component, args: [{
                        selector: 'ng-number-picker',
                        template: "\n  <div class=\"input-group mb-3 input-{{size}} {{customClass.container}}\">\n  <!-- Horizontal decrease button orientation -->\n  <div class=\"input-group-prepend\" *ngIf=\"isHorizontal() && showDownButton\">\n    <span class=\"input-group-text decrease {{customClass.down}}\" (click)=\"onDecrease($event)\"\n    (mouseup)=\"onMouseUp($event, false)\" (mousedown)=\"onMouseDown($event, false)\">-</span>\n  </div>\n  <!-- Input prefix -->\n  <div *ngIf=\"prefix\" class=\"input-group-prepend\">\n    <span class=\"input-group-text {{customClass.prefix}}\">{{prefix}}</span>\n  </div>\n  <input type=\"number\" class=\"form-control\" name=\"input-spin-val\"\n  [(ngModel)]=\"value\"\n  [readOnly]=\"inputReadOnly\"\n  (blur)=\"onBlur($event)\"\n  (focus)=\"onFocus($event)\"\n  (mousewheel)=\"mouseWheel && onMouseWheel($event)\"\n  (keyup)=\"arrowKeys && onKeyUp($event)\"\n  (keydown)=\"arrowKeys && onKeyDown($event)\"\n  (keydown.arrowup)=\"arrowKeys && onIncrease($event)\"\n  (keydown.arrowdown)=\"arrowKeys && onDecrease($event)\"\n  (change)=\"onValueChange($event)\"\n  [placeholder]=\"placeholder\"\n  >\n  <!-- Input postfix -->\n  <div *ngIf=\"postfix\" class=\"input-group-prepend\">\n    <span class=\"input-group-text {{customClass.postfix}}\" [style.borderLeft]=\"'0'\">{{postfix}}</span>\n  </div>\n  <!-- Horizontal increase button orientation -->\n  <div class=\"input-group-prepend\" *ngIf=\"isHorizontal() && showUpButton\">\n    <span class=\"input-group-text increase {{customClass.up}}\" [style.borderLeft]=\"(!postfix) ? '0' : ''\" (click)=\"onIncrease($event)\"\n    (mouseup)=\"onMouseUp($event)\" (mousedown)=\"onMouseDown($event)\">+</span>\n  </div>\n  <!-- Vertical buttons orientation -->\n  <div class=\"input-group-append\" *ngIf=\"!isHorizontal()\">\n    <span class=\"input-group-text vertical p-0\">\n      <span *ngIf=\"showUpButton\" class=\"{{customClass.up}}\" (click)=\"onIncrease($event)\" (mouseup)=\"onMouseUp($event)\"\n      (mousedown)=\"onMouseDown($event)\">+</span>\n      <span *ngIf=\"showDownButton\" class=\"{{customClass.down}}\" (click)=\"onDecrease($event)\" (mouseup)=\"onMouseUp($event, false)\"\n      (mousedown)=\"onMouseDown($event, false)\">-</span>\n    </span>\n  </div>\n</div>\n  ",
                        styles: ["input[type=number]::-webkit-inner-spin-button,input[type=number]::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}.decrease:hover,.increase:hover{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:pointer;background-color:#d8d8d8}.increase{border-top-right-radius:3px!important;border-bottom-right-radius:3px!important}.vertical{display:flex;justify-content:center;flex-direction:column;text-align:center;-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:pointer}.vertical span{line-height:15px}.input-md .vertical span,.input-sm .vertical span{padding:2px 10px}.input-lg .vertical span{padding:4px 10px}.input-xlg .vertical span{padding:7px 10px}.input-md,.input-medium{height:45px}.input-large,.input-lg{height:50px}.input-xlarge,.input-xlg{height:75px}.input-md input,.input-md span,.input-medium input,.input-medium span{font-size:22px}.input-large input,.input-large span,.input-lg input,.input-lg span{font-size:25px}.input-xlarge input,.input-xlarge span,.input-xlg input,.input-xlg span{font-size:38px}"]
                    },] },
        ];
        /** @nocollapse */
        NumberPickerComponent$$1.ctorParameters = function () {
            return [
                { type: NumberPickerService, },
            ];
        };
        NumberPickerComponent$$1.propDecorators = {
            "min": [{ type: core.Input },],
            "max": [{ type: core.Input },],
            "step": [{ type: core.Input },],
            "value": [{ type: core.Input },],
            "pickStartAfter": [{ type: core.Input },],
            "pickTimer": [{ type: core.Input },],
            "prefix": [{ type: core.Input },],
            "postfix": [{ type: core.Input },],
            "placeholder": [{ type: core.Input },],
            "buttonsOrientation": [{ type: core.Input },],
            "size": [{ type: core.Input },],
            "customClass": [{ type: core.Input },],
            "mouseWheel": [{ type: core.Input },],
            "arrowKeys": [{ type: core.Input },],
            "inputReadOnly": [{ type: core.Input },],
            "showUpButton": [{ type: core.Input },],
            "showDownButton": [{ type: core.Input },],
            "valueChange": [{ type: core.Output },],
            "minReached": [{ type: core.Output },],
            "maxReached": [{ type: core.Output },],
            "pickStarted": [{ type: core.Output },],
            "pickStoped": [{ type: core.Output },],
            "pickUpStarted": [{ type: core.Output },],
            "pickUpStoped": [{ type: core.Output },],
            "pickDownStarted": [{ type: core.Output },],
            "pickDownStoped": [{ type: core.Output },],
        };
        return NumberPickerComponent$$1;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var NumberPickerModule = (function () {
        function NumberPickerModule() {
        }
        NumberPickerModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            forms.FormsModule
                        ],
                        providers: [NumberPickerService],
                        declarations: [NumberPickerComponent$$1],
                        exports: [NumberPickerComponent$$1]
                    },] },
        ];
        return NumberPickerModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.NumberPickerService = NumberPickerService;
    exports.NumberPickerComponent = NumberPickerComponent$$1;
    exports.NumberPickerModule = NumberPickerModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
